### PR TITLE
Ruby: additional unsafe deserialization sinks for ox and one for oj

### PR DESCRIPTION
**Pull request by p-**
This PR adds additional unsafe deserialization sinks for the [ox](https://github.com/ohler55/ox) XML serialization library:
* `Ox.parse_obj`
* `Ox.load` (if mode is set to `:object`; mode can be specified as argument or globally.)

and one additional sink for the [oj](https://github.com/ohler55/oj) JSON serialization library:
* `Oj.object_load`

Since both libraries are from the same author they both have a similar option to set global deserialization options `Oj.default_options` respectively `Ox.default_options`. That's why I tried to refactor the existing `isOjModePair` and `OjOptionsHashWithModeKey` into `isModePair` and `OptionsHashWithModeKey` to reuse these parts of the query code. It might not have been the most elegant approach though 😅 .

*Important difference regarding global options*: while the global options of Oj are insecure by default, the global options of Ox are secure by default.

### CODE REVIEW

File: ruby/ql/lib/codeql/ruby/security/UnsafeDeserializationCustomizations.qll [hunk 1]
```diff
@@ -223,13 +223,109 @@ module UnsafeDeserialization {
...
         )
       )
     }
   }
 
+  /**
+   * The first argument in a call to `Oj.object_load`, always considered as a
+   * sink for unsafe deserialization. (global and local mode options are ignored)
+   */
+  class OjObjectLoadArgument extends Sink {
```
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/security/UnsafeDeserializationCustomizations.qll:237]: I think this class can be `private`.

```suggestion
  private class OjObjectLoadArgument extends Sink {
```

File: ruby/ql/lib/codeql/ruby/security/UnsafeDeserializationCustomizations.qll [hunk 2]
```diff
@@ -223,13 +223,109 @@ module UnsafeDeserialization {
...
+    SetOxDefaultOptionsCall() {
+      this = API::getTopLevelMember("Ox").getAMethodCall("default_options=")
+    }
+
+    /**
+     * Gets the value being assigned to `Ox.default_options`.
+     */
+    DataFlow::Node getValue() {
+      result.asExpr() =
+        this.getArgument(0).asExpr().(CfgNodes::ExprNodes::AssignExprCfgNode).getRhs()
```
Comment (aibaars) [inline ruby/ql/lib/codeql/ruby/security/UnsafeDeserializationCustomizations.qll:270]: Shouldn't this just be as follows? I believe the first argument of a `default_options=` function is the right-hand side of the assignment already. Same for the `Oj` variant of this class.

```suggestion
      result =  this.getArgument(0)
```

```suggestion
      result = this.getArgument(0)
```

**Review by aibaars — COMMENTED** [review]
Looks good to me. Only a few minor suggestions and questions.

File: ruby/ql/lib/codeql/ruby/security/UnsafeDeserializationCustomizations.qll (hunk 2 above)
Comment (p-) [reply ruby/ql/lib/codeql/ruby/security/UnsafeDeserializationCustomizations.qll:270]: Yes, now that you say it... makes a lot of sense 😄 

(FYI: I'll update this branch and check what fails)

**Review by aibaars — APPROVED** [review]
LGTM

**Review by aibaars — APPROVED** [review]

**Pull request merged**